Content.Client/UserInterface/Systems/Ghost/Controls/Roles/GhostRolesWindow.xaml.cs

```diff
@@ -12,20 +15,89 @@ public sealed partial class GhostRolesWindow : DefaultWindow
         public event Action<GhostRoleInfo>? OnRoleFollow;
 
+        private List<Collapsible> _collapsibleBoxes = new();
+
+        public GhostRolesWindow()
+        {
+            RobustXamlLoader.Load(this);
+        }
+
         public void ClearEntries()
         {
             NoRolesMessage.Visible = true;
             EntryContainer.DisposeAllChildren();
+            _collapsibleBoxes.Clear();
+        }
+
+        public Dictionary<int, bool> SaveCollapsibleBoxesStates()
+        {
+            var collapsibleStates = new Dictionary<int, bool>();
+            foreach (var collapsible in _collapsibleBoxes)
+            {
+                if (int.TryParse(collapsible.Name, out var collapsibleId))
+                {
+                    collapsibleStates[collapsibleId] = collapsible.BodyVisible;
+                }
+            }
+            return collapsibleStates;
+        }
+
+        public void RestoreCollapsibleBoxesStates(Dictionary<int, bool> collapsibleStates)
+        {
+            foreach (var collapsible in _collapsibleBoxes)
+            {
+                if (int.TryParse(collapsible.Name, out var collapsibleId) && collapsibleStates.TryGetValue(collapsibleId, out var isOpen))
+                {
+                    collapsible.BodyVisible = isOpen;
+                }
+            }
         }
 
         public void AddEntry(string name, string description, bool hasAccess, FormattedMessage? reason, IEnumerable<GhostRoleInfo> roles, SpriteSystem spriteSystem)
         {
             NoRolesMessage.Visible = false;
 
-            var entry = new GhostRolesEntry(name, description, hasAccess, reason, roles, spriteSystem);
-            entry.OnRoleSelected += OnRoleRequestButtonClicked;
-            entry.OnRoleFollow += OnRoleFollow;
-            EntryContainer.AddChild(entry);
+            var ghostRoleInfos = roles.ToList();
+            var rolesCount = ghostRoleInfos.Count;
+
+            var info = new GhostRoleInfoBox(name, description);
+            var buttons = new GhostRoleButtonsBox(hasAccess, reason, ghostRoleInfos, spriteSystem);
+            buttons.OnRoleSelected += OnRoleRequestButtonClicked;
+            buttons.OnRoleFollow += OnRoleFollow;
+
+            EntryContainer.AddChild(info);
+
+            if (rolesCount > 1)
+            {
+                var buttonHeading = new CollapsibleHeading(Loc.GetString("ghost-roles-window-available-button", ("rolesCount", rolesCount)));
+
+                buttonHeading.AddStyleClass(ContainerButton.StyleClassButton);
+                buttonHeading.Label.HorizontalAlignment = HAlignment.Center;
+                buttonHeading.Label.HorizontalExpand = true;
+
+                var body = new CollapsibleBody
+                {
+                    Margin = new Thickness(0, 10, 0, 0),
+                };
+
+                var uniqueId = $"{name}-{description}-{ghostRoleInfos.Select(r => r.Requirements)}".GetHashCode();
```

Comment:
This key should be the same grouping key construction used by `GhostRolesEui`, do not construct it separately as that may get out of sync.

Also, you're still using `GetHashCode()`. Stop that, it won't work. Just use the key value directly, and don't store the data on the collapsible's `Name` field, make `_collapsibleBoxes` into a dictionary or something so you can store the data properly.

Reply:
Got it. Will do

Reply:
Done in [d18efe2](https://github.com/space-wizards/space-station-14/pull/32717/commits/d18efe21e8068e503a6d1a86d4705c4f7c4c3cce)